### PR TITLE
Fixed JSON_BIGINT_AS_STRING for older servers

### PR DESCRIPTION
**Pull request by danbovey**
Getting the same error @itp-elliot-irwin in #11 

Added a fix from http://stackoverflow.com/a/27909889/1255333

### CODE REVIEW

**Comment by SykesCottagesDev** [comment]
@romanpitak Any chance of getting this merged?

**Comment by romanpitak** [comment]
Sorry, I forgot about it. I'll try to get it merged on Monday.

via CloudMagic Email [https://cloudmagic.com/k/d/mailapp?ct=pa&cv=8.5.38&pv=6.0&source=email_footer_2] On Fri, Jun 03, 2016 at 5:20 PM, SykesCottagesDev < notifications@github.com [notifications@github.com] > wrote:
@romanpitak [https://github.com/romanpitak] Any chance of getting this merged?

**Comment by CarwynNelson** [comment]
Thanks :) (I wrote the previous comment on the company account, hence the comment deletion)

**Comment by CarwynNelson** [comment]
It seems that applying this change manually to our fork has actually broken something.

With this fix (when trying to call GetCampaignById) I get the following error
```php
 #message: "Object of class stdClass could not be converted to string"
 #code: 0
 #file: "/var/www/project_name/vendor/romanpitak/dotmailer-api-v2-client/src/DataTypes/MagicArray.php"
 #line: 24
```

If I reverse the change everything seems to work just fine, not sure what is going on here?

File: src/DataTypes/MagicArray.php [hunk 1]
```diff
@@ -35,6 +35,29 @@ public function __construct($value = null)
...
+     * @link http://stackoverflow.com/a/27909889/1255333
+     * @param $input
+     */
+    private function json_decode_bigint_as_string($json, $assoc = false, $depth = 512) {
+        if (version_compare(PHP_VERSION, '5.4.0', '>=') && !(defined('JSON_C_VERSION') && PHP_INT_SIZE > 4)) {
+            return json_decode($json, $assoc, $depth);
+        } else {
+            $max_int_length = strlen((string) PHP_INT_MAX) - 1;
+            $json_without_bigints = preg_replace('/:\s*(-?\d{'.$max_int_length.',})/', ': "$1"', $json);
+            return json_decode($json_without_bigints);
```
Comment (CarwynNelson) [inline src/DataTypes/MagicArray.php]: `$assoc` needs to be added here as well, if they mismatch it causes errors.

File: src/DataTypes/MagicArray.php [hunk 2]
```diff
@@ -35,6 +35,29 @@ public function __construct($value = null)
...
+     * @link http://stackoverflow.com/a/27909889/1255333
+     * @param $input
+     */
+    private function json_decode_bigint_as_string($json, $assoc = false, $depth = 512) {
+        if (version_compare(PHP_VERSION, '5.4.0', '>=') && !(defined('JSON_C_VERSION') && PHP_INT_SIZE > 4)) {
+            return json_decode($json, $assoc, $depth);
+        } else {
+            $max_int_length = strlen((string) PHP_INT_MAX) - 1;
+            $json_without_bigints = preg_replace('/:\s*(-?\d{'.$max_int_length.',})/', ': "$1"', $json);
+            return json_decode($json_without_bigints, $assoc);
```
Comment (CarwynNelson) [inline src/DataTypes/MagicArray.php]: Thanks for making the `$assoc` change :) I've just been thinking that we should probably also add $depth to the fallback json_encode?

**Comment by danbovey** [comment]
Yeah of course! :smiley:

**Comment by CarwynNelson** [comment]
@romanpitak Does this look okay to be merged?

**Pull request merged**